exp/addons/internal/controllers/clusterresourceset_controller.go

```diff
@@ -266,98 +267,152 @@ func (r *ClusterResourceSetReconciler) ApplyClusterResourceSet(ctx context.Conte
 	errList := make([]error, 0)
 	resourceSetBinding := clusterResourceSetBinding.GetOrCreateBinding(clusterResourceSet)
 
-	// Iterate all resources and apply them to the cluster and update the resource status in the ClusterResourceSetBinding object.
-	for _, resource := range clusterResourceSet.Spec.Resources {
-		strategy := addonsv1.ClusterResourceSetStrategy(clusterResourceSet.Spec.Strategy)
+	// The behavior depends on what type of strategy has been applied.
 
-		// If resource is already applied successfully and clusterResourceSet mode is "ApplyOnce", continue. (No need to check hash changes here)
-		if resourceSetBinding.IsApplied(resource) && strategy == addonsv1.ClusterResourceSetStrategyApplyOnce {
-			continue
-		}
+	strategy := addonsv1.ClusterResourceSetStrategy(clusterResourceSet.Spec.Strategy)
 
-		unstructuredObj, err := r.getResource(ctx, resource, cluster.GetNamespace())
-		if err != nil {
-			if err == ErrSecretTypeNotSupported {
-				conditions.MarkFalse(clusterResourceSet, addonsv1.ResourcesAppliedCondition, addonsv1.WrongSecretTypeReason, clusterv1.ConditionSeverityWarning, err.Error())
-			} else {
-				conditions.MarkFalse(clusterResourceSet, addonsv1.ResourcesAppliedCondition, addonsv1.RetrievingResourceFailedReason, clusterv1.ConditionSeverityWarning, err.Error())
-
-				// Continue without adding the error to the aggregate if we can't find the resource.
-				if apierrors.IsNotFound(err) {
-					continue
-				}
-			}
-			errList = append(errList, err)
-			continue
-		}
+	switch strategy {
```

Comment:
The code for the two `strategy` is almost identical besides the check for `oldHash==newHash`

Can we pull this out into a shared function and call it from both cases?
https://github.com/mesosphere/cluster-api/blob/8b97e8123e2f8f2a547c223b7af66c29592b45c6/exp/addons/internal/controllers/clusterresourceset_controller.go#L284-L334

Reply:
This is a work in progress. I needed to split it to understand why the logic of the applyonce was doing things in certain order. It is quite different and it needs to be adjusted for the new strategies carefully, but yeah it should be possible to combine logic once i am able too make the tests work.

Reply:
This has been addressed, testing passes.